cmd/gen-ca/main.go

```diff
@@ -0,0 +1,378 @@
+	pubKeyPath := flag.String("publicKey", "", "Path to public key for certificate") // this could also be in the profile
+	issuerPath := flag.String("issuer", "", "Path to issuer cert if generating an intermediate")
+	outputPath := flag.String("output", "", "Path to store generated PEM certificate")
+	flag.Parse()
+
+	if *module == "" {
+		log.Fatal("--module is required")
+	}
+	if *label == "" {
+		log.Fatal("--label is required")
+	}
+	if *id == "" {
+		log.Fatal("--id is required")
+	}
+	if *profilePath == "" {
+		log.Fatal("--profile is required")
+	}
+	if *pubKeyPath == "" {
+		log.Fatal("--publicKey is required")
+	}
+	if *outputPath == "" {
+		log.Fatal("--output is required")
+	}
+
+	ctx, session, err := pkcs11helpers.Initialize(*module, *slot, *pin)
+	if err != nil {
+		log.Fatalf("Failed to setup session and PKCS#11 context: %s", err)
+	}
+	log.Println("Opened PKCS#11 session")
+
+	privKey, err := getKey(ctx, session, *label, *id)
+	if err != nil {
+		log.Fatalf("Failed to retrieve private key handle: %s", err)
+	}
+	log.Println("Retrieved private key handle")
+
+	profileBytes, err := ioutil.ReadFile(*profilePath)
+	if err != nil {
+		log.Fatalf("Failed to read certificate profile %q: %s", *profilePath, err)
+	}
+	var profile certProfile
+	err = json.Unmarshal(profileBytes, &profile)
+	if err != nil {
+		log.Fatalf("Failed to parse certificate profile: %s", err)
+	}
+
+	pubPEMBytes, err := ioutil.ReadFile(*pubKeyPath)
+	if err != nil {
+		log.Fatalf("Failed to read public key %q: %s", *pubKeyPath, err)
+	}
+	pubPEM, _ := pem.Decode(pubPEMBytes)
+	if pubPEM == nil {
+		log.Fatal("Failed to parse public key")
+	}
+	pub, err := x509.ParsePKIXPublicKey(pubPEM.Bytes)
+	if err != nil {
+		log.Fatalf("Failed to parse public key: %s", err)
+	}
+
+	certTmpl, err := constructCert(ctx, &profile, pubPEM.Bytes, session)
+	if err != nil {
+		log.Fatalf("Failed to construct certificate from profile: %s", err)
+	}
+	log.Println("Generated tbs certificate")
+	var issuer *x509.Certificate
+	if *issuerPath != "" {
+		// If generating an intermediate load the parent issuer and
+		// set the Authority Key Identifier in the template
+		issuerPEMBytes, err := ioutil.ReadFile(*issuerPath)
+		if err != nil {
+			log.Fatalf("Failed to read issuer certificate %q: %s", *issuerPath, err)
+		}
+		issuerPEM, _ := pem.Decode(issuerPEMBytes)
+		if issuerPEM == nil {
+			log.Fatal("Failed to parse issuer certificate PEM")
+		}
+		issuer, err = x509.ParseCertificate(issuerPEM.Bytes)
+		if err != nil {
+			log.Fatalf("Failed to parse issuer certificate: %s", err)
+		}
+		certTmpl.AuthorityKeyId = issuer.SubjectKeyId
+	} else {
+		issuer = certTmpl
+	}
+
+	// We don't pass a random io.Reader here as the PKCS#11 wrapper doesn't use it
+	certBytes, err := x509.CreateCertificate(nil, certTmpl, issuer, pub, privKey)
+	if err != nil {
+		log.Fatalf("Failed to create certificate: %s", err)
+	}
+	log.Println("Signed certificate")
```

Comment:
We should immediately write this signed certificate to disk and print it. Otherwise, errors below, like a parsing error, mean we forever lose track of this thing we signed.